File: backend/checkout-api/app/src/main/java/develcode/checkout/domain/shared/events/Event.java
package develcode.checkout.domain.shared.events;

import java.io.Serializable;
import java.util.UUID;

import com.fasterxml.jackson.annotation.JsonProperty;

public abstract class Event implements Serializable {

    @JsonProperty("id")
    private final String id;
    @JsonProperty("name")
    private final String name;
    @JsonProperty("payload")
    private Object payload;

    protected Event(String name) {
        this.id = UUID.randomUUID().toString();
        this.name = name;
        this.payload = null;
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public <T> T getPayload(Class<T> clazz) throws ClassCastException {
        return clazz.cast(payload);
    }

    public void setPayload(Object payload) {
        this.payload = payload;
    }
}

File: backend/payment-api/src/domain/events/event.ts
import { UUIDGenerator } from 'src/utils/uuid-generator';

export abstract class Event {
  private id: string;
  private name: string;
  private payload: any;

  protected constructor(name: string) {
    this.id = UUIDGenerator.generate();
    this.name = name;
    this.payload = null;
  }

  public getName(): string {
    return this.name;
  }

  public getPayload(): any {
    return this.payload;
  }

  public setPayload(payload: any): void {
    this.payload = payload;
  }
}
